scripts/west_commands/runners/nrfjprog.py

```diff
@@ -151,10 +163,24 @@ def do_run(self, command, **kwargs):
                 program_cmd
             ])
         else:
-            if self.family == 'NRF52':
+            if self.family == 'NRF51':
+                commands.append(program_cmd + ['--sectorerase'])
+            elif self.family == 'NRF52':
                 commands.append(program_cmd + ['--sectoranduicrerase'])
             else:
-                commands.append(program_cmd + ['--sectorerase'])
+                uicr = {
+                    'NRF53': ((0x00FF8000, 0x00FF8800),
+                              (0x01FF8000, 0x01FF8800)),
+                    'NRF91': ((0x00FF8000, 0x00FF8800),),
+                }[self.family]
+
+                if has_region(uicr, self.hex_):
+                    # Hex file has UICR contents.
+                    raise RuntimeError('The hex file contains data placed in '
+                                       'the UICR, which needs a full erase. '
+                                       'Run west flash again with --erase.')
```

Comment:
This effectively means that a user doing:
```
ninja flash
```
can only flash once, and afterwards the user will be presented with this message.
Is that acceptable ?

How about letting the build system decide, so that if user has a complete image (all partitions incl. UICR), then a full erase is performed.
If only a partial image, then fail with above message.

Small note: `Run west flash again` you don't know if use invoked `west flash` or `ninja flash` 😉

Reply:
> Small note: `Run west flash again` you don't know if use invoked `west flash` or `ninja flash` wink

This is a west command. The build system targets by the same name are just for backwards compatibility for the people who refuse to learn that they are using west ;). I think this is OK as-is

Reply:
> How about letting the build system decide, so that if user has a complete image (all partitions incl. UICR), then a full erase is performed.
> If only a partial image, then fail with above message.

I think it should also be possible to override this with --force.

Reply:
> The build system targets by the same name are just for backwards compatibility for the people who refuse to learn that they are using west ;). I think this is OK as-is

Not sure if that comment also covered this:
> ```
> ninja flash
> ```
> can only flash once, and afterwards the user will be presented with this message.
> Is that acceptable ?

As long as `ninja flash` exists and is NOT deprecated, then I find it wrong to be able to have a command that will work successfully once, and thereafter having to use another command for future flashing.

There is a difference in having to add an additional arguments to a command, and then request him to use another command (`ninja` --> `west`).

(is it time to dreprecate `ninja flash|debug|..` and friends ?)

Reply:
> can only flash once, and afterwards the user will be presented with this message.
> Is that acceptable ?

I think for this SoC that's the best we can do; since nrfjprog can't `--sectoranduicrerase` this one, `ninja flash` is stuck between two bad choices:

- always succeed but wipe out the bootloader
- work once after a chip erase, fail afterwards

Or am I misunderstanding?

Reply:
> Or am I misunderstanding?

If a user is building a complete image, bootloader, spm, and app, then we could do an erase all with the flashing.
The build system knows about partitions, so it should be able to check if everything is merged, or if space is allocated for a bootloader, but bootloader is not part of the image.

And / or check the content at address `0x0`, and if that is not empty, refuse to flash an image without a bootloader.

Unless i'm the one missing something.

Reply:
> `ninja flash` is stuck between two bad choices

Note that both are better than the current situation:

 - Silently (only a mild warning) leave the flash in the wrong state.

> The build system knows about partitions

That is out of scope for this particular PR IMO. I just want to get something in that is better than what is currently there.
We can create the final solution later.

Reply:
>     * work once after a chip erase, fail afterwards

Note that this is NOT what this PR does. This PR will make ninja flash fail, even the first time after chiperase, IF the hex file has UICR contents.

I think a hard error is the easiest for the user to deal with.

Reply:
> > ```
> > * work once after a chip erase, fail afterwards
> > ```
> 
> 
> Note that this is NOT what this PR does. This PR will make ninja flash fail, even the first time after chiperase, IF the hex file has UICR contents.

Ok, although I don't like the fact that `ninja flash` fails (as long as we have not deprecated that build target), then i'm happy that the command is consistent in it behavior.

And looking at the code again, I do see that this is indeed the case.
But this also mean a user cannot do a full erase (e.g. by using `nrfjprog`), and then a `west flash`, user would still need to use `west flash --erase`.

As long as `west erase` doesn't exists, user would need to resort directly to `nrfjprog` for erasing, in which case he could also program with `nrfjprog`.
I think this is acceptable for now.

> I think a hard error is the easiest for the user to deal with.

agree. As long as behavior is consistent.

> We can create the final solution later.

Agreed. I also think that this use case shows why flashing should be an independent tool (`west`, `nrfjprog`, `openocd`, `JLink`, etc) and NOT a build target.
(Hint: Let's work on deprecating `ninja/make flash` as we now have `west`)